fixing dropping of secondary cidr blocks on first apply of VPC

nishant221 opened the pull request:
Issue https://github.com/aws-controllers-k8s/community/issues/1816


Description of changes:
Added the fix for a bug where Secondary VPC CIDR blocks are getting dropped in the first apply of VPC resource.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

LikithaVemulapalli commented:
Hello @nishant221, we don't generally add code in `sdk.go` manually, you need to re-generate the controller. We follow the pattern of adding [templates/hooks](https://github.com/aws-controllers-k8s/ec2-controller/tree/main/templates/hooks/vpc ) to include code in `sdk`. You need to figure out if these changes needs to be added in the end or after create, and include them in the respective file. In general we will be calling these files in [`generator.yaml`](https://github.com/aws-controllers-k8s/ec2-controller/blob/5d460dbad24e44a11e80a693fbd61d498820476b/generator.yaml#L592) and when we run `make build-controller SERVICE=ec2` from code-generator, changes will be included in sdk file.

nishant221 commented:
@LikithaVemulapalli  Thanks for the feedback/review . Will make the required changes and update the PR.

nishant221 commented:
Hi @LikithaVemulapalli ,

As per the suggestion, tried making the required changes. Kindly have a look.
Thanks

LikithaVemulapalli commented on apis/v1alpha1/ack-generate-metadata.yaml:
```diff
@@ -1,8 +1,8 @@
+  build_hash: 6542e04cc357215c91df94e747e3d412fb26df97
   go_version: go1.19
-  version: v0.26.1
+  version: 6542e04-dirty
```
This version seems incorrect! There shouldn't be any dirty tagged along with the version, could you please revert your `GOPROXY` changes if made and try regenerating the code.

RedbackThomson replied:
I think the issue here might also be that you need to fetch the tags. Try `git fetch --tags` in the `code-generator` repository

nishant221 replied:
Regenerated the code. Now the `version` is not getting tagged with `-dirty` .

@RedbackThomson  Thanks for the suggestion. It was fixed by doing both `git fetch --all --tags` and `git update-index --refresh`.

LikithaVemulapalli reviewed (commented):
Could you also include a simple e2e test to validate this scenario, Thanks :)

RedbackThomson reviewed (commented):
This is excellent! Thank you for taking the time to work on this.

There are still 2 cases that I think we might need to support:
- What happens when a user updates the list of CIDR blocks? Do we need to remove the old associations and add the new ones?
- What happens when we delete the VPC altogether? Do we need to take any actions to delete the CIDR blocks before doing anything else?

nishant221 commented:
@RedbackThomson  Thanks for the suggestions/review .


> What happens when a user updates the list of CIDR blocks? Do we need to remove the old associations and add the new ones?

As described in the issue description (https://github.com/aws-controllers-k8s/community/issues/1816 ) , the functionality (or updates) are working fine with object reapply. Hence, I believe no changes are needed with updates.
 
> What happens when we delete the VPC altogether? Do we need to take any actions to delete the CIDR blocks before doing anything else?
Deleting the VPC does not require any additional action to be taken.

LikithaVemulapalli commented:
discussed offline, adding `ok-to-test` label to see if the tests pass

RedbackThomson commented:
I have a hunch that if someone removes this secondary CIDR block from the list, that it won't be properly disassociated. That is because we are never calling [`DisassociateVpcCidrBlock`](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DisassociateVpcCidrBlock) in the hook code update logic. 

Would you mind adding a test for this condition (after creating the resource, remove a CIDR block and check the resource with boto3)?

nishant221 commented:
@RedbackThomson I think [DisassociateVpcCidrBlock](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DisassociateVpcCidrBlock) is getting called [here](https://github.com/aws-controllers-k8s/ec2-controller/blob/main/pkg/resource/vpc/hooks.go#L182) via [syncCIDRBlocks](https://github.com/aws-controllers-k8s/ec2-controller/blob/main/pkg/resource/vpc/hooks.go#L256) .

The reason why it's not updating the AWS VPC resource is a small bug in  [computeStringPDifference function] (https://github.com/aws-controllers-k8s/ec2-controller/pull/139/files#diff-8e9c68bc9772730bc1c3837f6adeefebab8a5752e52088e9a7138ae123a52f3cR152). Added the fix the PR.

Also, added an additional test as suggested. E2E test seems to be passing. Kindly have a look.

Thanks

RedbackThomson commented on pkg/resource/vpc/hooks.go:
```diff
@@ -258,6 +259,50 @@ func (rm *resourceManager) customUpdateVPC(
 		}
 	}
 
+	// CidrBlockAssociationSet needs to be updated in the Status
```
See my comments below - but these new lines shouldn't be necessary if we re-use `syncCIDRBlocks`

nishant221 replied:
modification done as suggested.

RedbackThomson commented on pkg/resource/vpc/hooks.go:
```diff
@@ -428,3 +473,48 @@ func updateTagSpecificationsInCreateRequest(r *resource,
 	}
 }
+
+func (rm *resourceManager) attachSecondaryCidrBlocks(ctx context.Context, desired *resource,
```
This code is essentially duplicating the logic already inside `syncCIDRBlocks` - except it uses a `CreateVpcOutput` rather than the resource status. I believe `sdk_create_post_set_output` hook code can call `syncCIDRBlocks` directly. We do the same thing for [S3 `Bucket`](https://github.com/aws-controllers-k8s/s3-controller/blob/main/pkg/resource/bucket/hook.go#L57)

I do see that you are using the response from this SDK call to set the `CidrBlockAssociation` struct, though, and I think this is preferable to calling `DescribeVpcs` again (since it doesn't require additional API calls). We don't have this logic in the body of `syncCIDRBlocks`, so it would be great if you could move that into there.

nishant221 replied:
@RedbackThomson Updated the PR with the recommended changes.

RedbackThomson replied:
I don't see a call to `syncCIDRBlocks` in any code paths coming off create. Could you please add logic within `sdk_create_post_set_output`?

nishant221 replied:
I have already added the logic within `sdk_create_post_set_output` 
https://github.com/aws-controllers-k8s/ec2-controller/pull/139/files#diff-533c8b725d5ee9f8fc14e5e3f070d31296017f0f5c6cd966a319219076948c97R5

RedbackThomson replied:
Oh my bad, when I clicked around I didn't see it. Must have missed it

RedbackThomson commented:
/ok-to-test

RedbackThomson commented:
/lgtm
